libp2p/autotls/peeridauth.nim

```diff
@@ -0,0 +1,202 @@
+  except HttpError:
+    raise newException(PeerIDAuthError, "Failed to start PeerID Auth")
+
+  # www-authenticate
+  let wwwAuthenticate = authStartResponse.headers.getString("www-authenticate")
+  if wwwAuthenticate == "":
+    raise newException(PeerIDAuthError, "www-authenticate not present in response")
+  let challengeClient = extractField(wwwAuthenticate, "challenge-client")
+  var serverPublicKey: PublicKey
+  try:
+    serverPublicKey = PublicKey
+      .init(decode(extractField(wwwAuthenticate, "public-key")).toByteSeq)
+      .get()
+  except ValueError:
+    raise newException(PeerIDAuthError, "Could not decode public-key")
+  let opaque = extractField(wwwAuthenticate, "opaque")
+
+  let hostname = "registration." & AutoTLSDNSServer # registration.libp2p.direct
+  let clientPubKeyB64 = base64.encode(peerInfo.publicKey.getBytes().get(), safe = true)
+  let challengeServer = randomChallenge()
+  let sig = peerIdSign(peerInfo.privateKey, challengeClient, serverPublicKey, hostname)
+  let authHeader =
+    "libp2p-PeerID public-key=\"" & clientPubKeyB64 & "\"" & ", opaque=\"" & opaque &
+    "\"" & ", challenge-server=\"" & challengeServer & "\"" & ", sig=\"" & sig & "\""
+
+  # Authorization
+  var authorizationResponse: HttpClientResponseRef
+  try:
+    authorizationResponse = await HttpClientRequestRef
+    .post(
+      session,
+      url,
+      body = $payload,
+      headers = [
+        ("Content-Type", "application/json"),
+        ("User-Agent", "nim-libp2p"),
+        ("authorization", authHeader),
+      ],
+    )
+    .get()
+    .send()
+  except HttpError:
+    raise newException(PeerIDAuthError, "Failed to send Authorization for PeerID Auth")
+
+  # check server's signature
+  let serverSig =
+    extractField(authorizationResponse.headers.getString("authentication-info"), "sig")
+  if not checkSignature(
+    serverSig, serverPublicKey, challengeServer, peerInfo.publicKey, hostname
+  ):
+    raise newException(PeerIDAuthError, "Failed to validate server's signature")
+
+  # Bearer token
+  return (
+    extractField(
+      authorizationResponse.headers.getString("authentication-info"), "bearer"
+    ),
+    authorizationResponse,
+  )
+
+proc peerIdAuthSend*(
+    url: string,
+    session: HttpSessionRef,
+    peerInfo: PeerInfo,
+    payload: JsonNode,
+    bearerToken: Opt[string] = Opt.none(string),
+): Future[(string, HttpClientResponseRef)] {.
+    async: (raises: [AutoTLSError, PeerIDAuthError, CancelledError])
+.} =
+  if bearerToken.isNone:
+    return await peerIdAuthenticate(url, session, peerInfo, payload)
+
+  let authHeader = "libp2p-PeerID bearer=\"" & bearerToken.get & "\""
+  var response: HttpClientResponseRef
+  try:
+    response = await HttpClientRequestRef
+    .post(
+      session,
+      url,
+      body = $payload,
+      headers = [
+        ("Content-Type", "application/json"),
+        ("User-Agent", "nim-libp2p"),
+        ("authorization", authHeader),
+      ],
+    )
+    .get()
```

Comment:
It's possible for `post` to return an error, right?

Reply:
Yes! Added a check for status codes different than `200`

Reply:
Apparently it'll never return an error https://github.com/status-im/nim-chronos/blob/b55e2816eb45f698ddaca8d8473e401502562db2/chronos/apps/http/httpclient.nim#L1050